cirq-core/cirq/devices/noise_utils.py

```diff
@@ -0,0 +1,155 @@
+# Copyright 2021 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from dataclasses import dataclass
+from typing import TYPE_CHECKING, Sequence
+import warnings
+import numpy as np
+
+if TYPE_CHECKING:
+    import cirq
+
+
+# Tag for gates to which noise must be applied.
+PHYSICAL_GATE_TAG = 'physical_gate'
+
+
+@dataclass(frozen=True)
+class OpIdentifier:
+    """Identifies an operation by gate and (optionally) target qubits."""
+
+    gate_type: type
+    qubits: Sequence['cirq.Qid']
+
+    def __init__(self, gate_type: type, *qubits: 'cirq.Qid'):
+        object.__setattr__(self, 'gate_type', gate_type)
+        object.__setattr__(self, 'qubits', qubits)
+
+    def swapped(self):
+        return OpIdentifier(self.gate_type, *self.qubits[::-1])
+
+    def __str__(self):
+        return f'OpID({self.gate_type.__qualname__}{self.qubits})'
+
+    def __repr__(self) -> str:
+        fullname = f'{self.gate_type.__module__}.{self.gate_type.__qualname__}'
+        qubits = ', '.join(map(repr, self.qubits))
+        return f'cirq.devices.noise_utils.OpIdentifier({fullname}, {qubits})'
+
+
+# TODO: expose all from top-level cirq?
+def decay_constant_to_xeb_fidelity(decay_constant: float, num_qubits: int = 2) -> float:
+    """Calculates the XEB fidelity from the depolarization decay constant.
+
+    Args:
+        decay_constant: Depolarization decay constant
+        num_qubits: Number of qubits
+    """
+    N = 2 ** num_qubits
+    return 1 - ((1 - decay_constant) * (1 - 1 / N))
```

Comment:
is there not some xeb implementation somewhere in cirq? Would this not be more appropriate there?

Reply:
Similarly below.

Reply:
`cirq.experiments` has an XEB implementation, but I'd prefer to keep this here for two reasons:

- Due to Cirq import-order restrictions, files in `cirq.devices` cannot rely on `cirq.experiments` without a hack to avoid circular dependencies, which I'd prefer to avoid.
- The XEB implementation code is dedicated to the internal details of performing XEB. While these conversion methods are tangentially related, I'd prefer to keep them together rather than distribute them to each of the associated benchmarks.